File: benchmarks/spectral-norm.go
package main

import (
	"fmt"
	"math"
	"time"
)

func A(i, j int) float64 {
	return 1.0 / float64(((i+j)*(i+j+1))/2+i+1)
}

func multiplyAv(v, Atv []float64) {
	n := len(v)
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += A(i, j) * v[j]
		}
		Atv[i] = sum
	}
}

func multiplyAtv(v, Atv []float64) {
	n := len(v)
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += A(j, i) * v[j]
		}
		Atv[i] = sum
	}
}

func multiplyAtAv(v, u, AtAv []float64) {
	multiplyAv(v, u)
	multiplyAtv(u, AtAv)
}

func main() {
	n := 1000
	t0 := time.Now()
	
	u := make([]float64, n)
	v := make([]float64, n)
	tmp := make([]float64, n)
	
	for i := 0; i < n; i++ {
		u[i] = 1.0
	}
	
	for i := 0; i < 10; i++ {
		multiplyAtAv(u, tmp, v)
		multiplyAtAv(v, tmp, u)
	}
	
	vBv := 0.0
	vv := 0.0
	for i := 0; i < n; i++ {
		vBv += u[i] * v[i]
		vv += v[i] * v[i]
	}
	
	res := math.Sqrt(vBv / vv)
	t1 := time.Now()
	
	fmt.Printf("%.9f\n", res)
	fmt.Printf("bench_ns=%d\n", t1.Sub(t0).Nanoseconds())
}

File: benchmarks/spectral-norm.rs
use std::time::Instant;

fn a(i: usize, j: usize) -> f64 {
    1.0 / (((i + j) * (i + j + 1) / 2 + i + 1) as f64)
}

fn multiply_av(v: &[f64], atv: &mut [f64]) {
    for i in 0..v.len() {
        let mut sum = 0.0;
        for j in 0..v.len() {
            sum += a(i, j) * v[j];
        }
        atv[i] = sum;
    }
}

fn multiply_atv(v: &[f64], atv: &mut [f64]) {
    for i in 0..v.len() {
        let mut sum = 0.0;
        for j in 0..v.len() {
            sum += a(j, i) * v[j];
        }
        atv[i] = sum;
    }
}

fn multiply_at_av(v: &[f64], tmp: &mut [f64], at_av: &mut [f64]) {
    multiply_av(v, tmp);
    multiply_atv(tmp, at_av);
}

fn main() {
    let n = 1000;
    let t0 = Instant::now();
    
    let mut u = vec![1.0; n];
    let mut v = vec![0.0; n];
    let mut tmp = vec![0.0; n];
    
    for _ in 0..10 {
        multiply_at_av(&u, &mut tmp, &mut v);
        multiply_at_av(&v, &mut tmp, &mut u);
    }
    
    let mut v_bv = 0.0;
    let mut vv = 0.0;
    for i in 0..n {
        v_bv += u[i] * v[i];
        vv += v[i] * v[i];
    }
    
    let res = (v_bv / vv).sqrt();
    let duration = t0.elapsed();
    
    println!("{:.9}", res);
    println!("bench_ns={}", duration.as_nanos());
}
